src/VisualStudio/Next/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs

```diff
@@ -0,0 +1,178 @@
+
+            private CancellationTokenSource _shutdown;
+            private Task<RemoteHostClient> _instance;
+
+            public RemoteHostClientService(Workspace workspace, IDiagnosticAnalyzerService analyzerService)
+            {
+                _gate = new object();
+
+                _workspace = workspace;
+                _analyzerService = analyzerService;
+            }
+
+            public void Enable()
+            {
+                lock (_gate)
+                {
+                    if (_instance != null)
+                    {
+                        // already enabled
+                        return;
+                    }
+
+                    if (!_workspace.Options.GetOption(RemoteHostOptions.RemoteHost))
+                    {
+                        // not turned on
+                        return;
+                    }
+
+                    // make sure we run it on background thread
+                    _shutdown = new CancellationTokenSource();
+
+                    var token = _shutdown.Token;
+                    _instance = Task.Run(() => EnableAsync(token), token);
+                }
+            }
+
+            public void Disable()
+            {
+                lock (_gate)
+                {
+                    if (_instance == null)
+                    {
+                        // already disabled
+                        return;
+                    }
+
+                    var instance = _instance;
+                    _instance = null;
+
+                    RemoveGlobalAssets();
+
+                    _shutdown.Cancel();
+
+                    try
+                    {
+                        instance.Result.Shutdown();
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // _instance wasn't finished running yet.
+                    }
+                }
+            }
+
+            public async Task<RemoteHostClient> GetRemoteHostClientAsync(CancellationToken cancellationToken)
+            {
+                // copy instance to local variable so that we don't need lock here
+                var instance = _instance;
+                if (instance == null)
+                {
+                    // service is in shutdown mode or not enabled
+                    return null;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // make sure instance was started
+                return await instance.ConfigureAwait(false);
+            }
+
+            private async Task<RemoteHostClient> EnableAsync(CancellationToken cancellationToken)
+            {
+                await AddGlobalAssetsAsync(cancellationToken).ConfigureAwait(false);
+
+                return await StartInternalAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<RemoteHostClient> StartInternalAsync(CancellationToken cancellationToken)
```

Comment:
used when user click re-enable from info bar. if we don't show it, then we can remove it